server/chat-commands/moderation.ts

```diff
@@ -110,76 +92,103 @@ export const commands: ChatCommands = {
 		const nextGroupName = nextGroup.name || "regular user";
 
-		if (currentSymbol === nextSymbol) {
-			return this.errorReply(`User '${name}' is already a ${nextGroupName} in this room.`);
-		}
-		if (!user.can('makeroom')) {
-			if (currentGroup.id && !user.can(`room${currentGroup.id || 'voice'}` as 'roomvoice', null, room)) {
-				return this.errorReply(`/${cmd} - Access denied for promoting/demoting from ${currentGroupName}.`);
+		for (const toPromote of users) {
+			const targetUser = Users.get(toPromote);
+			const userid = toID(toPromote);
+			let name = targetUser ? targetUser.name : this.filter(this.targetUsername);
+			if (!name) continue;
+			name = name.slice(0, 18);
+
+			if (!userid) return this.parse('/help roompromote');
+			if (!targetUser && !Users.isUsernameKnown(userid) && !force) {
+				this.errorReply(`User '${name}' is offline and unrecognized, and so can't be promoted.`);
+				continue;
 			}
-			if (nextSymbol !== ' ' && !user.can(`room${nextGroup.id || 'voice'}` as 'roomvoice', null, room)) {
-				return this.errorReply(`/${cmd} - Access denied for promoting/demoting to ${nextGroupName}.`);
+			if (targetUser && !targetUser.registered) {
+				this.errorReply(`User '${name}' is unregistered, and so can't be promoted.`);
+				continue;
 			}
-		}
-		if (targetUser?.locked && room.persist && room.settings.isPrivate !== true && nextGroup.rank > 2) {
-			return this.errorReply("Locked users can't be promoted.");
-		}
 
-		if (nextSymbol === Users.Auth.defaultSymbol()) {
-			room.auth.delete(userid);
-		} else {
-			room.auth.set(userid, nextSymbol);
-		}
+			let currentSymbol: GroupSymbol | 'whitelist' = room.auth.getDirect(userid);
+			if (room.auth.has(userid) && currentSymbol === Users.Auth.defaultSymbol()) {
+				currentSymbol = 'whitelist';
+			}
+			const currentGroup = Users.Auth.getGroup(currentSymbol);
+			const currentGroupName = currentGroup.name || "regular user";
 
-		// Only show popup if: user is online and in the room, the room is public, and not a groupchat or a battle.
-		const shouldPopup = (
-			targetUser && room.users[targetUser.id] && room.persist && room.settings.isPrivate !== true ?
-				targetUser : null
-		);
+			if (currentSymbol === nextSymbol) {
+				this.errorReply(`User '${name}' is already a ${nextGroupName} in this room.`);
+				continue;
+			}
+			if (!user.can('makeroom')) {
+				if (currentGroup.id && !user.can(`room${currentGroup.id || 'voice'}` as 'roomvoice', null, room)) {
+					this.errorReply(`/${cmd} - Access denied for promoting/demoting from ${currentGroupName}.`);
+					continue;
+				}
+				if (nextSymbol !== ' ' && !user.can(`room${nextGroup.id || 'voice'}` as 'roomvoice', null, room)) {
+					this.errorReply(`/${cmd} - Access denied for promoting/demoting to ${nextGroupName}.`);
+					continue;
+				}
+			}
+			if (targetUser?.locked && room.persist && room.settings.isPrivate !== true && nextGroup.rank > 2) {
+				this.errorReply("Locked users can't be promoted.");
```

Comment:
These three error messages (access denied promoting to, access denied promoting from, and locked users) happen every loop and so should mention the username.